feat(new): warn about potential duplicate patient (#2187)

janmarkusmilan opened the pull request:
Fixes #2187

**Changes proposed in this pull request:**

- A modal appears when a patient with matching name, sex and date of birth is created
- The modal outputs the message "Potential duplicate of: [Patient Name](link to patient record). Are you sure you want to create this patient?"
- Clicking on the patient name will redirect to the patient view
- Clicking on the cancel button will close the modal
- Clicking on the continue button will dispatch create patient, redirect to the new patient view and output a success message
- Added a new test file for the modal

**No new dependency added**

jackcmeyer commented on src/patients/new/NewPatient.tsx:
```diff
@@ -41,13 +45,39 @@ const NewPatient = () => {
   }
 
   const onSave = () => {
-    dispatch(createPatient(patient, onSuccessfulSave))
+    let isDuplicatePatient = false
+    const patientsObj = {}
+    Object.assign(patientsObj, patients)
+    Object.keys(patientsObj).forEach((patientInfo: any) => {
+      const loggedPatient = patients[patientInfo]
+      if (
+        loggedPatient.givenName === patient.givenName &&
+        loggedPatient.familyName === patient.familyName &&
+        loggedPatient.sex === patient.sex &&
+        loggedPatient.dateOfBirth === patient.dateOfBirth
+      ) {
+        setShowDuplicateNewPatientModal(true)
+        setDuplicatePatient(loggedPatient as Patient)
+        isDuplicatePatient = true
+      }
+    })
+    if (!isDuplicatePatient) {
+      dispatch(createPatient(patient, onSuccessfulSave))
+    }
   }
```
After adding support for finding all possible duplicate patients, rather than just the first, this function would be a good candidate for `Arrays.filter` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter. 


```ts
const onSave = () => {
	 const duplicatePatients = patients.filter((existingPatient) => isPossibleDuplicatePatient(patient, existingPatient));
	
	if(duplicatePatients.length > 0) {
       setShowDuplicateNewPatientModal(true)
       setDuplicatePatient(existingPatient)
	} else {
      dispatch(createPatient(patient, onSuccessfulSave))
	}
}
```

jackcmeyer commented on src/patients/new/NewPatient.tsx:
```diff
@@ -10,6 +10,7 @@ import Patient from '../../shared/model/Patient'
 import { RootState } from '../../shared/store'
 import GeneralInformation from '../GeneralInformation'
 import { createPatient } from '../patient-slice'
+import DuplicateNewPatientModal from './DuplicateNewPatientModal'
```
There should be some tests to test this new workflow. 

Basically, have an already existing patient in the redux store, try to create a new patient with the same information, and test to make sure that the new modal shows.

jackcmeyer commented on src/__tests__/patients/new/DuplicateNewPatientModal.test.tsx:
```diff
@@ -0,0 +1,107 @@
+
+  describe('on save', () => {
+    it('should call the onContinueButtonClick function when the continue button is clicked', () => {
+      const onContinueButtonClickSpy = jest.fn()
+      const store = mockStore({
+        patient: {
+          patient: {
+            id: '1234',
+          },
+        },
+      } as any)
+
+      const wrapper = mount(
+        <Provider store={store}>
+          <DuplicateNewPatientModal
+            show
+            toggle={jest.fn()}
+            onCloseButtonClick={jest.fn()}
+            onContinueButtonClick={onContinueButtonClickSpy}
+          />
+        </Provider>,
+      )
+      wrapper.update()
```
This code block is duplicated several times in this file. I think we can extract it to a setup function that looks like: 


```ts
const setup = () => {
 const onContinueButtonClickSpy = jest.fn()
      const store = mockStore({
        patient: {
          patient: {
            id: '1234',
          },
        },
      } as any)

      const wrapper = mount(
        <Provider store={store}>
          <DuplicateNewPatientModal
            show
            toggle={jest.fn()}
            onCloseButtonClick={jest.fn()}
            onContinueButtonClick={onContinueButtonClickSpy}
          />
        </Provider>,
      )
	wrapper.update()

   return { wrapper: wrapper as ReactWrapper }
}
```

janmarkusmilan commented:
Sorry, I accidentally committed again (did not make the changes yet). Working on it right now.

janmarkusmilan commented on src/patients/new/DuplicateNewPatientModal.tsx:
```diff
@@ -0,0 +1,54 @@
+  const body = (
+    <div className="row">
+      <div className="col-md-12">
+        <p>
+          Possible duplicate of:{' '}
+          {duplicatePatient !== undefined && (
+            <Link to={`/patients/${duplicatePatient.id}`}>{duplicatePatient.fullName}</Link>
+          )}
+        </p>
+        <p>Are you sure you want to create this patient?</p>
```
Just for clarification, do you mean using the translator? For example:

    <Alert color="danger" message={t('addTheTextToBeInternationalizedHere')} />

janmarkusmilan commented on src/patients/new/NewPatient.tsx:
```diff
@@ -21,8 +22,11 @@ const NewPatient = () => {
   const history = useHistory()
   const dispatch = useDispatch()
   const { createError } = useSelector((state: RootState) => state.patient)
+  const { patients } = Object(useSelector((state: RootState) => state.patients))
```
I originally had that line, but the tests would always fail. The message is:

    TypeError: Cannot destructure property 'patients' of '(0 , _reactRedux.useSelector)(...)' as it is undefined.

The only solution I found was adding the Object(...) at the beginning.

janmarkusmilan commented:
# Recently Pushed Major Changes

## 1. Helper util file (is-possible-duplicate-patient.ts)

![is-possible-duplicate-patient.ts](https://i.imgur.com/dwSSi5P.png)

Created helper util file to allow for reusability if necessary.

---

## 2. Support for finding multiple duplicate patients

### NewPatient.tsx

![NewPatient.tsx](https://i.imgur.com/3kaJEnR.png)

### DuplicateNewPatientModal.tsx

![DuplicateNewPatientModal.tsx](https://i.imgur.com/kkT7hqX.png)

Used the `Arrays.filter` method and the newly created helper util file `is-possible-duplicate-patient.ts` to loop through the existing patients and determine if there are zero, one or multiple patients with the same information as the one being created. Also styled the look of the modal based on if there is one or multiple patients.

---

## 3. New alert and internationalized text

### DuplicateNewPatientModal.tsx

![Updated_Alert](https://i.imgur.com/nI59GXK.png)

Added an alert to bring more attention to the warning being provided. Also, all the text is internationalized (still missing english translation key in console).

### Modal (Multiple Patients)

![Multiple_Patients](https://i.imgur.com/giJeZjU.png)

### Modal (Single Patient)

![Single_Patient](https://i.imgur.com/26ZfJuR.png)

---

## 4. Setup function in test file

### DuplicateNewPatientModal.test.tsx

![DuplicateNewPatientModal.text.tsx](https://i.imgur.com/KGqVsXu.png)

### Using the setup function

![Code_Reusability](https://i.imgur.com/JcwmFL5.png)

Since this code block was duplicated multiple times, it was extracted to a setup function for easy reusability.

---

## 5. Test case for new modal

### NewPatient.test.tsx

![NewPatient.test.tsx](https://i.imgur.com/cL0uLGt.png)

### NewPatient.tsx

![Test_Patient](https://i.imgur.com/Jz3ahzD.png)

A test needed to be created in order to test the new modal feature. Given some existing patient in the redux store, if a newly created patient had the same information as the existing one, then the modal should render (or in this case return true).  

---

# Other Things

1.  Changed `const [duplicatePatient, setDuplicatePatient] = useState({} as Patient)` to `const [duplicatePatient, setDuplicatePatient] = useState<Patient | undefined>(undefined)`
2. Had to keep `const { patients } = Object(useSelector((state: RootState) => state.patients))` as it is, since removing the `Object(...)` results in errors when running the test cases for `NewPatient.tsx`

jackcmeyer commented on src/patients/new/DuplicateNewPatientModal.tsx:
```diff
@@ -0,0 +1,68 @@
+  const { duplicatePatient, show, toggle, onCloseButtonClick, onContinueButtonClick } = props
+
+  const alertMessage =
+    'Patient with matching information found in database. Are you sure you want to create this patient?'
```
this alert message needs to have the string internationalized.

jackcmeyer commented on src/patients/new/DuplicateNewPatientModal.tsx:
```diff
@@ -0,0 +1,68 @@
+      <Alert color="danger" title={t('Warning!')} message={t(alertMessage)} />
+      <div className="row">
+        <div className="col-md-12">
+          {`${t('Possible duplicate patient')}: `}
```
The value passed to be a key that is defined in the `locales` folder.

jackcmeyer commented on src/patients/new/DuplicateNewPatientModal.tsx:
```diff
@@ -0,0 +1,68 @@
+
+  const body = (
+    <>
+      <Alert color="danger" title={t('Warning!')} message={t(alertMessage)} />
```
Thoughts on having this as a warning color rather than danger?

janmarkusmilan replied:
I think it makes sense. It's more of a warning message than an error message, and it still stands out.

jackcmeyer commented on src/patients/util/is-possible-duplicate-patient.ts:
```diff
@@ -0,0 +1,13 @@
+import Patient from '../../shared/model/Patient'
+
+export function isPossibleDuplicatePatient(newPatient: Patient, existingPatient: Patient) {
```
as trivial as it is, we should have a test around this functionality.

jackcmeyer replied:
I would expect a function named `isPossibleDuplicatePatient` to return a boolean rather than a object. 

I think this function can be simplified as: 

```ts
export function isPossibleDuplicatePatient(newPatient: Patient, existingPatient: Patient) {
	return newPatient.givenName === existingPatient.givenName &&
    newPatient.familyName === existingPatient.familyName &&
    newPatient.sex === existingPatient.sex &&
    newPatient.dateOfBirth === existingPatient.dateOfBirth
}

```

jackcmeyer commented on src/patients/new/DuplicateNewPatientModal.tsx:
```diff
@@ -0,0 +1,68 @@
+    <>
+      <Alert color="danger" title={t('Warning!')} message={t(alertMessage)} />
+      <div className="row">
+        <div className="col-md-12">
+          {`${t('Possible duplicate patient')}: `}
+          {duplicatePatient !== undefined &&
+            Object.entries(duplicatePatient).length === 1 &&
+            Object.entries(duplicatePatient).map(([key, patient]) => (
+              <Link key={key.toString()} to={`/patients/${patient.id}`}>
+                {patient.fullName}
+              </Link>
+            ))}
+          {duplicatePatient !== undefined &&
+            Object.entries(duplicatePatient).length > 1 &&
+            Object.entries(duplicatePatient).map(([key, patient]) => (
+              <li key={key.toString()}>
+                <Link to={`/patients/${patient.id}`}>{patient.fullName}</Link>
+              </li>
+            ))}
+        </div>
```
I think that it is ok if it always displays a list no matter the number of duplicate patients returned. If we do that, then we can simplify this code.

janmarkusmilan commented:
## Newly Pushed Changes

### 1. Changed alert color and internationalized text

I changed the color of the alert to "warning" instead of danger, and internationalized the text by adding them in the `locales` folder.

---

### 2. Changed `is-possible-duplicate-patient.ts` to return a boolean instead of an object, and created test file

I simplified the util file to return a boolean instead of an object, and created a test file for the function.

---

### 3. Set modal to display a list regardless of the amount of duplicate patients

I simplified the code by always rendering a list of duplicate patients in the modal.

---

![Modal](https://i.imgur.com/zmeuufg.png)
